cannon/mipsevm/multithreaded/testutil/expectations.go

```diff
@@ -0,0 +1,203 @@
+		HI:               fromThread.Cpu.HI,
+		LO:               fromThread.Cpu.LO,
+		Registers:        fromThread.Registers,
+		Dropped:          false,
+	}
+}
+
+func (e *ExpectedMTState) ExpectStep() {
+	// Set some standard expectations for a normal step
+	e.Step += 1
+	e.PrestateActiveThread().PC += 4
+	e.PrestateActiveThread().NextPC += 4
+	e.StepsSinceLastContextSwitch += 1
+}
+
+func (e *ExpectedMTState) ExpectPreemption(preState *multithreaded.State) {
+	e.ActiveThreadId = FindNextThread(preState).ThreadId
+	e.StepsSinceLastContextSwitch = 0
+	if preState.TraverseRight {
+		e.TraverseRight = e.RightStackSize > 1
+		e.RightStackSize -= 1
+		e.LeftStackSize += 1
+	} else {
+		e.TraverseRight = e.LeftStackSize == 1
+		e.LeftStackSize -= 1
+		e.RightStackSize += 1
+	}
+}
+
+func (e *ExpectedMTState) ExpectNewThread() *ExpectedThreadState {
+	newThreadId := e.NextThreadId
+	e.NextThreadId += 1
+	e.ThreadCount += 1
+
+	// Clone expectations from prestate active thread's original state (bf changing any expectations)
+	newThread := &ExpectedThreadState{}
+	*newThread = e.prestateActiveThreadOrig
+
+	newThread.ThreadId = newThreadId
+	e.threadExpectations[newThreadId] = newThread
+
+	return newThread
+}
+
+func (e *ExpectedMTState) ActiveThread() *ExpectedThreadState {
+	return e.threadExpectations[e.ActiveThreadId]
+}
+
+func (e *ExpectedMTState) PrestateActiveThread() *ExpectedThreadState {
+	return e.threadExpectations[e.prestateActiveThreadId]
+}
+
+func (e *ExpectedMTState) Thread(threadId uint32) *ExpectedThreadState {
+	return e.threadExpectations[threadId]
+}
+
+func (e *ExpectedMTState) Validate(t require.TestingT, actualState *multithreaded.State) {
+	require.Equal(t, e.PreimageKey, actualState.GetPreimageKey(), fmt.Sprintf("Expect preimageKey = %v", e.PreimageKey))
+	require.Equal(t, e.PreimageOffset, actualState.GetPreimageOffset(), fmt.Sprintf("Expect preimageOffset = %v", e.PreimageOffset))
+	require.Equal(t, e.Heap, actualState.GetHeap(), fmt.Sprintf("Expect heap = 0x%x", e.Heap))
+	require.Equal(t, e.ExitCode, actualState.GetExitCode(), fmt.Sprintf("Expect exitCode = 0x%x", e.ExitCode))
+	require.Equal(t, e.Exited, actualState.GetExited(), fmt.Sprintf("Expect exited = %v", e.Exited))
+	require.Equal(t, e.Step, actualState.GetStep(), fmt.Sprintf("Expect step = %d", e.Step))
+	require.Equal(t, e.LastHint, actualState.GetLastHint(), fmt.Sprintf("Expect lastHint = %v", e.LastHint))
+	require.Equal(t, e.MemoryRoot, common.Hash(actualState.GetMemory().MerkleRoot()), fmt.Sprintf("Expect memory root = %v", e.MemoryRoot))
+	// Thread-related global fields
+	require.Equal(t, e.StepsSinceLastContextSwitch, actualState.StepsSinceLastContextSwitch, fmt.Sprintf("Expect StepsSinceLastContextSwitch = %v", e.StepsSinceLastContextSwitch))
+	require.Equal(t, e.Wakeup, actualState.Wakeup, fmt.Sprintf("Expect Wakeup = %v", e.Wakeup))
+	require.Equal(t, e.TraverseRight, actualState.TraverseRight, fmt.Sprintf("Expect TraverseRight = %v", e.TraverseRight))
```

Comment:
You can use `require.Equalf` to remove the need for the `fmt.Sprintf`

https://pkg.go.dev/github.com/stretchr/testify/require#Equalf

Reply:
awesome - thanks!